### ISSUE TITLE
Ambiguity in Server Side Integration and Problems with Serialization for #[server] macro

### ISSUE THREAD

**Issue by MinaMatta98** [leptos-rs/leptos#813]
Hello,

### TLDR ###

* Is it intended by design that basic REST API integration be done via the server or via leptos's #[server] api. I ask as this will likely affect reactivity, although I can attach an on-response hook just fine.

* In terms of #[server] requiring that a component be serialized, are there any additional constraints which the docs have not mentioned, or am I running into a bug at runtime?

I had a rather long question as to how this library was to interact within a global server environment. I would appreciate some guidance as the examples do not cover this specific problem.

Let's take a post request for example, I could go down the standard actix path in serializing a file if I wanted:

```
#[post("/submit")]
pub async fn upload_server(mut form_data: Payload) -> Result<String, std::io::Error> {
    use futures_util::StreamExt;

    let mut bytes = BytesMut::new();
    while let Some(item) = form_data.next().await {
        bytes.extend_from_slice(&item.unwrap());
    }

    std::fs::write("./upload/file.pdf", bytes).unwrap();
    Ok("hello".to_string())}
```

There is also the #[server] macro provided by leptos, however, it isn't very clear what the initial scope of the use case for this was. I know the macro can be used to extract responses via [`ResponseOptions`](https://docs.rs/leptos_actix/latest/leptos_actix/struct.ResponseOptions.html)  , fairly easily. However, It isn't very clear how it's meant to interact with forms.

For example, I have assumed that [web-sys](https://rustwasm.github.io/wasm-bindgen/web-sys/index.html#the-web-sys-crate) is needed for form-extraction. I have attached the following callback to the form:

```

    let on_submit_callback = move |event: SubmitEvent| {
        event.stop_propagation();
        event.prevent_default();

        let form = event.target().unwrap().unchecked_into::<HtmlFormElement>();

        let file_input = form
            .elements()
            .named_item("file-input")
            .unwrap()
            .unchecked_into::<HtmlInputElement>();

        let file = file_input.files().unwrap().get(0).unwrap();
        let buffer = file_to_u8(file);
        spawn_local(async move {
            let buffer = buffer.await.unwrap();

            let mut vec_u8 = vec![0u8; buffer.length() as usize];
            buffer.copy_to(&mut vec_u8);
            log!("{:?}", vec_u8);
            file_upload(vec_u8).await.unwrap();
            toggle.set(false);
        });
    };

```

I basically call an external function and have it serialize into a [js_sys](https://docs.rs/js-sys/latest/js_sys/index.html)::[Uint8Array](https://docs.rs/js-sys/latest/js_sys/struct.Uint8Array.html#), transform it into a Vec<u8> and push that to the following server function:
```
#[server(FileUpload, "/upload")]
async fn file_upload(array: Vec<u8>) -> Result<String, ServerFnError> {
    std::fs::write("./upload/file.pdf", array).unwrap();
    Ok("hello".to_string())
}
```

Now, I get the following error:

![image](https://user-images.githubusercontent.com/81748812/230253130-20de208b-af9a-4f85-8341-2bb72607d08c.png)

It's unclear to me whether this is a bug in the #[server] macro, as there does not seem to be a problem with compilation, but instead serialization at runtime. I have had a brief look at the vector and all bytes seem to be within the 0-255, so I doubt it's an encoding issue.

Your help is appreciated, but not expected. Cheers.

**Comment by gbj** [comment]
It's somewhat hard to see which `.unwrap()` is panicking in your code because I don't have access to the line numbers, but I imagine `src/pages/augment/mod.rs:179` might be this line?
```rust
file_upload(vec_u8).await.unwrap();
```
The error `unsupported value` would be coming from `serde_urlencoded` in that case, which possibly doesn't support serializing `Vec<u8>`.

An alternative approach would be to use the `CBOR` encoding support by changing your server function to 
```rust
#[server(FileUpload, "/upload", "Cbor")]
```

For file uploads in particular, it's also possible simply to use a normal server API endpoint as you note, as multipart form data/file uploads can be tricky to handle well (although it seems you're doing it well already).

It's important to say that the framework has no opinion on whether you use `#[server]` functions or a REST API or GraphQL or anything else to integrate with the server... It's up to you and what works best for your use case and your own preferences. You can create an entire application with no server functions, or with only server functions, or with a mixture, and any of that is totally fine.

I'm going to close the issue because I don't think there's an underlying bug here, but feel free to reply, as I'll still see it.

**Comment by MinaMatta98** [comment]
Hello,

Thanks for this. 

I believe that it is slightly easier using standard actix API's for file uploads and have gone with that. 

It may be advantageous to support [actix-multipart](https://crates.io/crates/actix-multipart) in the future within the `#[server]` macro, although this is a creature comfort.

I will just leave this for anyone who wants to know how to transform a [js-sys::promise](https://docs.rs/js-sys/0.3.61/js_sys/struct.Promise.html) to a Future. You can deal with submitting a post request via web-sys using the following example, which also demonstrates how to retrieve the response:
```
            buffer.copy_to(&mut vec_u8);

            let request = Request::new_with_str_and_init(
                "/submit",
                RequestInit::new().method("POST").body(Some(&buffer)),
            )
            .unwrap();

            let window = web_sys::window().unwrap();

            let result = JsFuture::from(window.fetch_with_request(&request))
                .await
                .unwrap()
                .dyn_into::<web_sys::Response>()
                .unwrap();
            if result.ok() {
                response_setter.set(view! {cx,
                    <div style="display: block ruby" >
                          <p class="mx-1">"File Uploaded Successfully "</p>
                          <i class="fas fa-check-circle text-success" style="font-size: 1em"></i>
                    </div>
                });
                toggle.set(false);
            } else {
                let response_text = JsFuture::from(result.text().unwrap())
                    .await
                    .unwrap()
                    .as_string()
                    .unwrap();
                response_setter.set(view! {cx,
                <div style="display: block ruby">
                      <p class="mx-1">{response_text}</p>
                      <i class="fas fa-circle-xmark text-danger" style="font-size: 1em"></i>
                </div>});
            }
```

**Comment by gbj** [comment]
For your benefit and for others, I would add that the [`gloo_net`](https://docs.rs/gloo-net/latest/gloo_net/) crate provides wrappers that make it much easier to work with fetch requests like this, rather than manually working with `JsFuture`s and so on.
